sdks/java/core/src/main/java/org/apache/beam/sdk/util/TimerInternals.java

```diff
@@ -50,17 +50,36 @@
 public interface TimerInternals {
 
   /**
-   * Writes out a timer to be fired when the watermark reaches the given
-   * timestamp.
+   * Writes out a timer to be fired when the current time in the specified time domain reaches the
+   * target timestamp.
    *
-   * <p>The combination of {@code namespace}, {@code timestamp} and {@code domain} uniquely
-   * identify a timer. Multiple timers set for the same parameters can be safely deduplicated.
+   * <p>The combination of {@code namespace} and {@code timerId} uniquely identify a timer.
+   *
+   * <p>If a timer is set and then set again before it fires, later settings should clear the prior
+   * setting.
+   *
+   * <p>It is an error to set a timer for two different time domains.
+   */
+  void setTimer(StateNamespace namespace, String timerId, Instant target, TimeDomain timeDomain);
```

Comment:
I'm inclined to remove `setTimer(TimerData)` entirely. Any place that doesn't care about the ID can just use a dummy ID.

Or, failing that, I might change it to `setWakeup(StateNamespace, Instant, TimeDomain)` or some such. But I think the extra method doesn't add much value.